src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,795 @@
+            .unwrap()
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Reclock offsets 3 and 4 to timestamp 2000
+        batch.insert(
+            PART_ID,
+            vec![(3, MzOffset::from(3)), (4, MzOffset::from(4))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .unwrap()
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Compact enough so that we can correctly timestamp only offsets >= 3
+        operator.compact(Antichain::from_elem(1000)).await;
+
+        // Reclock offsets 3 and 4 again to see we haven't lost the ability
+        batch.insert(
+            PART_ID,
+            vec![(3, MzOffset::from(3)), (4, MzOffset::from(4))],
+        );
+        let reclocked_msgs = operator
+            .reclock(&mut batch)
+            .await
+            .unwrap()
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Attempt to reclock offset 2 should return an error
+        batch.insert(PART_ID, vec![(2, MzOffset::from(2))]);
+        assert_eq!(
+            operator.reclock(&mut batch).await.and(Ok(())),
+            Err((PART_ID, 2.into()))
+        );
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn test_concurrency() {
+        const PART_ID: PartitionId = PartitionId::None;
+
+        // Create two operators pointing to the same shard
+        let shared_shard = ShardId::new();
+        let location = PersistLocation {
+            blob_uri: "mem://".to_owned(),
+            consensus_uri: "mem://".to_owned(),
+        };
+        let client = location.open().await.unwrap();
+        let mut op_a = test_operator_with_client(client.clone(), shared_shard).await;
+        let mut op_b = test_operator_with_client(client, shared_shard).await;
+
+        // Reclock a batch from one of the operators
+        let mut batch = HashMap::new();
+
+        tokio::time::advance(Duration::from_secs(1)).await;
+
+        // Reclock offsets 1 and 2 to timestamp 1000 from operator A
+        batch.insert(
+            PART_ID,
+            vec![(1, MzOffset::from(1)), (2, MzOffset::from(2))],
+        );
+        let reclocked_msgs = op_a
+            .reclock(&mut batch)
+            .await
+            .unwrap()
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Also compact operator A. Since operator B has its own read handle it shouldn't affect it
+        op_a.compact(Antichain::from_elem(1000)).await;
```

Comment:
can you additionally call `reclock_frontier` on `op_a` here? It doesnt mutate the operator, but we can ensure that A is making progress as expected